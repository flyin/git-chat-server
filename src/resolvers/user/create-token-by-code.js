const mongoose = require('mongoose');
const { get, find } = require('lodash');
const axios = require('axios');
const GitHubApi = require('github');

const settings = require('../../settings');


const User = mongoose.model('User');

module.exports = async (_, { code }) => {
  throw new Error('Not implemented yet');

  const response = await axios({
    data: {
      client_id: settings.github.clientId,
      client_secret: settings.github.clientSecret,
      code
    },

    headers: {
      'Accept': 'application/json'
    },

    method: 'post',
    url: 'https://github.com/login/oauth/access_token'
  });

  if (!response.data.access_token) {
    throw new Error(get(response.data, 'error_description', 'oauth_receive_error'));
  }

  const client = createGithubClient();
  client.authenticate({ token: response.data.access_token, type: 'oauth' });
  const userResponse = await client.users.get({});
  let emailsResponse;

  if (!userResponse.data.email) {
    emailsResponse = await client.users.getEmails({ page: 1 });
  }

  return User.findOneAndUpdate(
    { email: userResponse.data.email || (find(emailsResponse.data, { primary: true, verified: true }) || {}).email },

    {
      avatar: userResponse.data.avatar_url,

      github: {
        accessToken: response.data.access_token,
        githubId: userResponse.data.id,
        name: userResponse.data.name,
        scopes: userResponse.data.scopes
      }
    },

    { new: true, upsert: true }
  );
};

function createGithubClient() {
  return new GitHubApi({ debug: !settings.isProduction, headers: { 'user-agent': '@flyin/git-chat-server' } });
}

