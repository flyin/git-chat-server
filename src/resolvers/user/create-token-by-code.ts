import signUser from './helpers/sign-user';
import { User } from 'models';
import { GraphQLContext } from 'services';
import { githubCreateClient, githubGetPrimaryEmail, githubGetToken, githubGetUser } from 'helpers';

type Input = {
  code: string
};

export default async (_: any, { code }: Input, context: GraphQLContext) => {
  const oauth = await githubGetToken(code);

  if (!oauth.access_token) {
    throw new Error('access_token_error');
  }

  const client = githubCreateClient(oauth.access_token);
  const githubUser = await githubGetUser(client);
  const primaryEmail = await githubGetPrimaryEmail(client);
  const currentUser = await context.currentUser;

  const user = await User.findOneAndUpdate(
    { $or: [{ _id: currentUser ? currentUser._id : null }, { 'github.githubId': githubUser.id }, { email: primaryEmail.email }] },

    {
      $setOnInsert: {
        email: primaryEmail.email
      },

      $set: {
        github: {
          avatar: githubUser.avatar_url,
          accessToken: oauth.access_token,
          githubId: githubUser.id,
          name: githubUser.name,
          scope: oauth.scope
        }
      }
    },

    { new: true, upsert: true }
  );

  if (!user) {
    throw new Error('user_find_and_update_error');
  }

  return signUser(user);
};
