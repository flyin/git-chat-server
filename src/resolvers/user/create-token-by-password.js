const mongoose = require('mongoose');

const signUser = require('./helpers/sign-user');

const User = mongoose.model('User');

module.exports = async (_, { email, password }) => {
  throw new Error('Not implemented yet');

  const user = await User.findOne({ email });

  if (!user || !await user.passwordIsValid(password)) {
    throw new Error('Login or password incorrect')
  }

  return signUser(user);
};
