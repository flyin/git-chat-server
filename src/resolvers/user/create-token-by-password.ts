import { User } from 'models';
import signUser from './helpers/sign-user';

export default async (_: any, { email, password }: { email: string, password: string }) => {
  const user = await User.findOne({ email });

  if (!user || !user.id || !await user.passwordIsValid(password)) {
    throw new Error('Login or password incorrect');
  }

  return signUser(user);
};
