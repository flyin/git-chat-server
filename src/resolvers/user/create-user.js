const mongoose = require('mongoose');


const User = mongoose.model('User');

module.exports = async (user) => {
  throw new Error('Not implemented yet');

  return await User.create(user);
};
