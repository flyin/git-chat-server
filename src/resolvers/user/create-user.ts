import { User } from 'models';

export default async (_: any, user: { email: string, password: string }) => {
  return User.create(user);
};
