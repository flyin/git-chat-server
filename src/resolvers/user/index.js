module.exports = {
  Mutation: {
    createTokenByCode: require('./create-token-by-code'),
    createTokenByPassword: require('./create-token-by-password'),
    createUser: require('./create-user')
  },

  Query: {
    getUser: require('./get-user')
  }
};
