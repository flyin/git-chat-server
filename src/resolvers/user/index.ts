import createUser from './create-user';
import createTokenByPassword from './create-token-by-password';
import createTokenByCode from './create-token-by-code';
import getUser from './get-user';

export default {
  Mutation: {
    createTokenByCode,
    createTokenByPassword,
    createUser
  },

  Query: {
    getUser
  }
};
