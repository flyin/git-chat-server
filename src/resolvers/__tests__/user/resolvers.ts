import { get } from 'lodash';
import { graphql } from 'graphql';
import { initTestCase } from 'utils/tests-utils';
import schema from 'schema';
import userResolvers from 'resolvers/user';
import { mongoose } from 'services/mongoose';
import { UserModel } from 'models/user';

jest.mock('axios');
jest.mock('github');
initTestCase();

const axios = require('axios');

// TODO implement mock setup for response
const Github = require('github');

const { createUser } = userResolvers.Mutation;
const userModel = mongoose.model<UserModel>('User');

describe('user resolvers', () => {
  describe('mutation', () => {
    beforeEach(() => {
      axios._setupMock([
        {
          url: 'https://github.com/login/oauth/access_token',
          method: 'post',
          data: { access_token: 'some-token', scope: 'user:email,admin:repo_hook' }
        }
      ]);

      Github._setupMock([
        {
          data: { id: 1234, name: 'flyin', avatar: 'http://avatar.example.com' },
          entity: 'users',
          method: 'get',
          resolve: true
        },

        {
          data: [
            { email: 'wrong@example.com', verified: false, primary: false },
            { email: 'verified@example.com', verified: true, primary: false },
            { email: 'primary-verified@example.com', verified: true, primary: true },
            { email: 'another-verified@example.com', verified: true, primary: false }
          ],

          entity: 'users',
          method: 'getEmails',
          resolve: true
        }
      ]);
    });

    afterEach(() => {
      axios._clean();
      Github._clean();
    });

    const userCreateQuery = `
      mutation user($email: String!, $password: String!) {
        createUser(email: $email, password: $password) {
          email
          isAdmin
        }
      }
    `;

    it('user can register with email and password', async () => {
      const user = { email: 'a@b.ru', password: '123' };
      const result = await graphql(schema, userCreateQuery, {}, {}, user);
      expect(result.data).toMatchSnapshot();
      expect(result.errors).toBeUndefined();
    });

    it('user cant register with invalid data', async () => {
      const user = { email: 'a@', password: '123' };
      const result = await graphql(schema, userCreateQuery, {}, {}, user);
      expect(result.data).toBeNull();
      expect(result.errors).toMatchSnapshot();
    });

    it('user can exchange email and password for token', async () => {
      const user = { email: 'aa@aa.ru', password: '123' };
      await createUser(null, user);

      const query = `
        mutation token($email: String!, $password: String!) {
          createTokenByPassword(email: $email, password: $password) {
            token
            iat
            exp
            user {
              email
            }
          }
        }
      `;

      const result = await graphql(schema, query, {}, {}, user);
      const returnData = get(result.data, 'createTokenByPassword', {});

      expect(result.errors).toBeUndefined();
      expect(returnData).toHaveProperty('token');
      expect(returnData).toHaveProperty('user');
      expect(get(returnData, 'user')).toMatchObject({ email: user.email });
    });

    it('user can register with github token', async () => {
      const query = `
        mutation token($code: String!) {
          createTokenByCode(code: $code) {
            token
            user {
              email
              isAdmin
              github {
                githubId
                name
                avatar
              }
            }
          }
        }
      `;

      const result = await graphql(schema, query, {}, {}, { code: 'some-github-code' });
      const returnData = get(result.data, 'createTokenByCode', {});

      expect(result.errors).toBeUndefined();
      expect(returnData).toHaveProperty('token');
      expect(returnData).toHaveProperty('user');
      expect(get(returnData, 'user')).toMatchSnapshot();
    });

    it('user can attach github to his account', async () => {
      const user = await createUser(null, { email: 'flyin@example.com', password: '123123' });

      const query = `
        mutation token($code: String!) {
          createTokenByCode(code: $code) {
            token
            user {
              email
              isAdmin
              github {
                githubId
                name
                avatar
              }
            }
          }
        }
      `;

      const result = await graphql(schema, query, {}, { currentUser: Promise.resolve(user) }, { code: 'some-github-code' });
      const returnData = get(result.data, 'createTokenByCode', {});
      expect(result.errors).toBeUndefined();
      expect(returnData).toHaveProperty('token');
      expect(returnData).toHaveProperty('user');
      expect(get(returnData, 'user')).toMatchSnapshot();
    });

    it('user can disconnect github from account');
    it('user can reset their password');
  });

  describe('query', () => {
    const query = `
        query user($userId: ID) {
          getUser(userId: $userId) {
            email
            isAdmin
          }
        }
      `;

    it('user can access to own profile', async () => {
      let result;
      const user = await createUser(null, { email: 'aa@aa.ru', password: '123' });
      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(user) });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(user) }, { userId: 'some-exists-or-not-id' });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);
    });

    it('admin user can access to any profile', async () => {
      let result;
      const user = await createUser(null, { email: 'aa@aa.ru', password: '123' });

      const adminUser = await userModel.create({
        email: 'admin@example.com',
        password: '123',
        isAdmin: true
      });

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(adminUser) }, { userId: user.id });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(user.email);

      result = await graphql(schema, query, {}, { currentUser: Promise.resolve(adminUser) }, { userId: adminUser.id });

      expect(result.errors).toBeUndefined();
      expect(get(result.data, 'getUser.email')).toBe(adminUser.email);
    });
  });

  describe('subscriptions', () => {
    it('notify socket then user is registered');
  });
});
