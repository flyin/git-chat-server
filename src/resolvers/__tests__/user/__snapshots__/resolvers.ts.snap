// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resolvers mutation user can attach github to his account 1`] = `
Object {
  "email": "flyin@example.com",
  "github": Object {
    "avatar": null,
    "githubId": "1234",
    "name": "flyin",
  },
  "isAdmin": false,
}
`;

exports[`user resolvers mutation user can register with email and password 1`] = `
Object {
  "createUser": Object {
    "email": "a@b.ru",
    "isAdmin": false,
  },
}
`;

exports[`user resolvers mutation user can register with github token 1`] = `
Object {
  "email": "primary-verified@example.com",
  "github": Object {
    "avatar": null,
    "githubId": "1234",
    "name": "flyin",
  },
  "isAdmin": false,
}
`;

exports[`user resolvers mutation user cant register with invalid data 1`] = `
Array [
  [GraphQLError: User validation failed: email: Email address \`a@\` is incorrect],
]
`;
